gapis/api/vulkan/vulkan.go

```diff
@@ -78,6 +78,137 @@ func (s *State) SetupInitialState(ctx context.Context, state *api.GlobalState) {
 	}
 }
 
+// TrimInitialState scans the capture commands to see which parts of the initial
+// state are actually used, and removes some unused parts from it.
+//
+// Note: the current approach consists in "manually" monitoring which Vulkan
+// objects are being used in callbacks passed to sync.MutateWithSubcommands,
+// however this basically re-encode some state tracking logic found in the API
+// files. A better way would be to use an api.StateWatcher and rely on api.RefID
+// to track which objects are accessed: this would avoid to re-encode state
+// tracking logic here. There might be some pitfalls though, e.g. when a command
+// just reads the handle of an object, the state watcher would not mark an
+// access to that object. For instance, when creating a derivate pipeline, a
+// VkPipeline handle is used in BasePipelineHandle, the API implementation reads
+// this handle, but does not access the corresponding object. So using
+// api.StateWatcher might need some wider design considerations.
+func (s *State) TrimInitialState(ctx context.Context, capturePath *path.Capture) error {
+	// Parts of the state we want to record the usage of.
+	descriptorSets := map[VkDescriptorSet]struct{}{}
+	pipelines := map[VkPipeline]struct{}{}
+
+	// Record usage in initial state.
+	for _, ci := range s.LastComputeInfos().All() {
+		pipelines[ci.ComputePipeline().VulkanHandle()] = struct{}{}
+	}
+	for _, di := range s.LastDrawInfos().All() {
+		for _, d := range di.DescriptorSets().All() {
+			descriptorSets[d.VulkanHandle()] = struct{}{}
+		}
+		pipelines[di.GraphicsPipeline().VulkanHandle()] = struct{}{}
+	}
+
+	// Record usage in the trace commands
+	// top-level commands
+	postCmdCb := func(s *api.GlobalState, subCmdIdx api.SubCmdIdx, cmd api.Cmd) {
+		switch cmd := cmd.(type) {
+		case *VkFreeDescriptorSets:
+			count := cmd.DescriptorSetCount()
+			ds, err := cmd.PDescriptorSets().Slice(0, (uint64)(count), s.MemoryLayout).Read(ctx, cmd, s, nil)
+			if err != nil {
+				panic("Read error")
```

Comment:
Can you panic the error, instead. Or use `log.Errf` and panic its returned error.

Reply:
done